File: skills/code-architect/references/code-examples/swift/Mediator.swift
// Mediator — centralize communication between components.

protocol Mediator: AnyObject {
    func notify(sender: Component, event: String)
}

class Component {
    weak var mediator: Mediator?
    let id: String
    init(id: String) { self.id = id }
}

final class Publisher: Component {
    func publish(_ event: String) {
        print("\(id) publishes '\(event)'")
        mediator?.notify(sender: self, event: event)
    }
}

final class Subscriber: Component {
    func receive(_ event: String) {
        print("\(id) received '\(event)'")
    }
}

// Mediator wires senders to receivers; components stay decoupled.
final class EventHub: Mediator {
    private var subscribers: [Subscriber] = []

    func register(_ subscriber: Subscriber) {
        subscriber.mediator = self
        subscribers.append(subscriber)
    }

    func register(_ publisher: Publisher) {
        publisher.mediator = self
    }

    func notify(sender: Component, event: String) {
        guard sender is Publisher else { return }
        subscribers.forEach { $0.receive(event) }
    }
}

func runDemo() {
    let hub = EventHub()
    let publisher = Publisher(id: "P1")
    hub.register(publisher)
    hub.register(Subscriber(id: "S1"))
    hub.register(Subscriber(id: "S2"))

    publisher.publish("config-changed")
}

runDemo()

File: skills/code-architect/references/code-examples/typescript/Mediator.ts
// Mediator — centralize communication between components.

interface Mediator {
  notify(sender: Component, event: string): void;
}

abstract class Component {
  // Mediator is attached after construction to avoid a circular dependency.
  protected mediator: Mediator | null = null;
  setMediator(mediator: Mediator): void {
    this.mediator = mediator;
  }
}

class TextInput extends Component {
  private text = "";
  type(value: string): void {
    this.text = value;
    this.mediator?.notify(this, "input");
  }
  get value(): string {
    return this.text;
  }
}

class SubmitButton extends Component {
  enabled = false;
  click(): void {
    if (this.enabled) {
      this.mediator?.notify(this, "submit");
    }
  }
}

// Concrete mediator wires components without them knowing each other.
class FormMediator implements Mediator {
  constructor(private input: TextInput, private button: SubmitButton) {
    input.setMediator(this);
    button.setMediator(this);
  }

  notify(sender: Component, event: string): void {
    if (sender === this.input && event === "input") {
      this.button.enabled = this.input.value.length > 0;
      console.log(`button enabled: ${this.button.enabled}`);
    }
    if (sender === this.button && event === "submit") {
      console.log(`submitting "${this.input.value}"`);
    }
  }
}

function demo(): void {
  const input = new TextInput();
  const button = new SubmitButton();
  new FormMediator(input, button);

  input.type("hello");
  button.click();
}

demo();
